ledger/accountdb.go

```diff
@@ -633,43 +1055,741 @@ func accountsAddNormalizedBalance(tx *sql.Tx, proto config.ConsensusParams) erro
+		return err
+	}
+	for rows.Next() {
+		var addrbuf []byte
+		var encodedAcctData []byte
+		var normBal sql.NullInt64
+		err = rows.Scan(&addrbuf, &encodedAcctData, &normBal)
+		if err != nil {
+			return err
+		}
+
+		var accountData basics.AccountData
+		err = protocol.Decode(encodedAcctData, &accountData)
+		if err != nil {
+			return err
+		}
+		var newAccountData baseAccountData
+		newAccountData.SetAccountData(&accountData)
+		encodedAcctData = protocol.Encode(&newAccountData)
+
+		if normBal.Valid {
+			insertRes, err = insertNewAcctBaseNormBal.ExecContext(ctx, addrbuf, encodedAcctData, normBal.Int64)
+		} else {
+			insertRes, err = insertNewAcctBase.ExecContext(ctx, addrbuf, encodedAcctData)
+		}
+
+		if err != nil {
+			return err
+		}
+		rowsAffected, err = insertRes.RowsAffected()
+		if err != nil {
+			return err
+		}
+		if rowsAffected != 1 {
+			return fmt.Errorf("number of affected rows is not 1 - %d", rowsAffected)
+		}
+		rowID, err = insertRes.LastInsertId()
+		if err != nil {
+			return err
+		}
+		insertResourceCallback := func(ctx context.Context, rowID int64, cidx basics.CreatableIndex, rd *resourcesData) error {
+			var err error
+			if rd != nil {
+				encodedData := protocol.Encode(rd)
+				_, err = insertResources.ExecContext(ctx, rowID, cidx, encodedData)
+			}
+			return err
+		}
+		err = accountDataResources(ctx, &accountData, rowID, insertResourceCallback)
+		if err != nil {
+			return err
+		}
+		processedAccounts++
+		if log != nil {
+			log(processedAccounts, totalBaseAccounts)
+		}
+	}
+
+	// if the above loop was abrupt by an error, test it now.
+	if err = rows.Err(); err != nil {
+		return err
+	}
+
+	for _, stmt := range applyNewAcctBase {
+		_, err = tx.Exec(stmt)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// removeEmptyAccountData removes empty AccountData msgp-encoded entries from accountbase table
+// and optionally returns list of addresses that were eliminated
+func removeEmptyAccountData(tx *sql.Tx, queryAddresses bool) (num int64, addresses []basics.Address, err error) {
+	if queryAddresses {
+		rows, err := tx.Query("SELECT address FROM accountbase where length(data) = 1 and data = x'80'") // empty AccountData is 0x80
+		if err != nil {
+			return 0, nil, err
+		}
+		defer rows.Close()
+
+		for rows.Next() {
+			var addrbuf []byte
+			err = rows.Scan(&addrbuf)
+			if err != nil {
+				return 0, nil, err
+			}
+			var addr basics.Address
+			if len(addrbuf) != len(addr) {
+				err = fmt.Errorf("account DB address length mismatch: %d != %d", len(addrbuf), len(addr))
+				return 0, nil, err
+			}
+			copy(addr[:], addrbuf)
+			addresses = append(addresses, addr)
+		}
+
+		// if the above loop was abrupted by an error, test it now.
+		if err = rows.Err(); err != nil {
+			return 0, nil, err
+		}
+	}
+
+	result, err := tx.Exec("DELETE from accountbase where length(data) = 1 and data = x'80'")
+	if err != nil {
+		return 0, nil, err
+	}
+	num, err = result.RowsAffected()
```

Comment:
if we ignore the affected result, why not skip this check?

Reply:
that's probably a good point, but I'd rather leaving it as is for now. This code is part of the upgradeDatabaseSchema4, and it wasn't causing any issue I'm aware of. If we were to change this, I would want to have some testing around this change - which I'd like to avoid for now..